flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java

```diff
@@ -354,22 +354,27 @@ public void merge(W mergeResult,
 					}
 				});
 
-				// drop if the window is already late
-				if (isLate(actualWindow)) {
-					mergingWindows.retireWindow(actualWindow);
-					continue;
-				}
+				context.key = key;
+				context.window = actualWindow;
 
 				W stateWindow = mergingWindows.getStateWindow(actualWindow);
 				if (stateWindow == null) {
 					throw new IllegalStateException("Window " + window + " is not in in-flight window set.");
 				}
 
 				windowState.setCurrentNamespace(stateWindow);
-				windowState.add(element.getValue());
 
-				context.key = key;
-				context.window = actualWindow;
+				// Drop if the window is already late. In rare cases (with a misbehaving
+				// WindowAssigner) it can happen that a window becomes late that already has
+				// state (contents, state and timers). That's why we first get the window state
+				// above and then drop everything.
+				if (isLate(actualWindow)) {
+					clearAllState(actualWindow, windowState, mergingWindows);
+					mergingWindows.persist();
```

Comment:
Yes，in fact，I had seen that, but I did not realize that it could be deleted.Haha, You are very quick-witted.:)
Anyway, I must say thanks for your explaining.
Best,
SunJincheng